src/persist/src/nemesis/mod.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Invariant-based random tests for the persist crate.
+//!
+//! This module generates random traffic against the external API of the persist
+//! crate. It records the requests sent and responses received and afterward
+//! checks this history against the API invariants.
+//!
+//! API traffic is modeled as a series of [Input]s. These are created by a
+//! [Generator], which is totally deterministic given a seed. The [Input]s are
+//! handed to an implementation of [Runtime] which executes the embedded [Req]
+//! and saves the [Res] wrapped up with metadata about the execution into a
+//! [Step]. These steps collectively form a history, which is handed to
+//! [Validator] to check for API invariant violations (representing a bug in the
+//! persist crate).
+//!
+//! The initial Runtime implementation is [Direct].
+//!
+//! Several of these components can be configured. Particularly interesting
+//! combinations of these are committed as unit tests. By default, these unit
+//! tests all select and use a random seed for traffic generation. This runs the
+//! test randomly once and is of limited usefulness.
+//!
+//! ```shell
+//! cargo test -p persist -- nemesis
+//! ```
+//!
+//! The seed is printed when the test is run and override-able, so if a test
+//! does fail (persist violated one of its invariants aka a bug), the previous
+//! traffic can be exactly replayed. For a deterministic runtime (the "direct_*"
+//! variants) this results in exactly the same execution sequence and can be
+//! used to verify the bug fix.
+//!
+//! ```shell
+//! NEMESIS_SEED=16385522461931935384 cargo test -p persist -- failing_test_name
+//! ```
+//!
+//! If the runtime is not deterministic, the seed can still be fixed to increase
+//! the likelihood of ticking the same bug again. The cargo-stress crate, which
+//! runs a unit test (or set of tests) in a tight loop until failure, is useful
+//! for this.
+//!
+//! ```shell
+//! NEMESIS_SEED=16385522461931935384 cargo stress -p persist -- failing_test_name
+//! ```
+//!
+//! Finally, the cargo-stress crate is also useful without fixing the seed to
+//! search for new invariant violations. This repeatedly runs all nemesis tests,
+//! each with new random traffic to maximize the chances of finding an
+//! interesting history.
+//!
+//! ```shell
+//! cargo stress -p persist -- nemesis
+//! ```
+
+// TODO
+// - Variants with FileBuffer and FileBlob
```

Comment:
we'll also want to add support for destroying streams once we build that out

Reply:
Added